In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [3]:
dataset = pd.read_csv('eie_full_translated.csv')

/var/folders/cz/wq9d8j_11fx3b8pjjbk7z8br0000gn/T/ipykernel_93373/3111726938.py:1: DtypeWarning: Columns (74,75,77,78,79,80,88,89,91,92,93,94,95,96,98,99,100,101,105,106,119) have mixed types. Specify dtype option on import or set low_memory=False.
  dataset = pd.read_csv('eie_full_translated.csv')


In [4]:
dataset.head()

outid sextypename                regname   
0  a99c6c63-aa70-4aec-ba42-370f7261e857        Male    Zaporizhzhia Oblast  \
1  c3136421-569e-422e-ae8f-41c4c931fd70      Female    Khmelnytskyi Oblast   
2  fbd0e280-4b77-4f34-942d-00bc850c2992      Female  Dnipropetrovsk Oblast   
3  30de395e-7a74-452a-8370-6856d240fbfb        Male      Chernivtsi Oblast   
4  52a40c74-374e-4744-bd6f-938ab7a4a641        Male        Mykolaiv Oblast   

                 areaname                           tername   
0      Melitopol district                       s. Patience  \
1   Krasylivskyi district              the city of Krasiliv   
2  Petropavlovsk district                 Dmytrivka village   
3              Chernivtsi  Shevchenkiv district of the city   
4    Vradiivskyi district                    Kumari village   

                                         regtypename   
0  A graduate of an Ukrainian school of the curre...  \
1  A graduate of an Ukrainian school of the curre...   
2       A graduate of a school of the previous years   
3  A graduate of an Ukrainian school of the curre...   
4  College student of a vocational or higher educ...   

                                              eoname   
0  Terpinniv Collegium "Zherelo" of the Melitopol...  \
1  Krasyliv secondary school of grades I-III No. ...   
2                                                NaN   
3  Chernivtsi specialized school of the I-III deg...   
4                                                NaN   

                           eotypename            eoregname   
0                           Collegium  Zaporizhzhia Oblast  \
1  Secondary general education school  Khmelnytskyi Oblast   
2                                 NaN                  NaN   
3                  Specialized school    Chernivtsi Oblast   
4                                 NaN                  NaN   

              eoareaname  ... ukradaptscale Degree of institution   
0     Melitopol district  ...           NaN                   NaN  \
1  Krasylivskyi district  ...           NaN                   NaN   
2                    NaN  ...           NaN                   NaN   
3             Chernivtsi  ...           NaN                   NaN   
4                    NaN  ...           NaN                   NaN   

  Number of full-time teaching staff in total   
0                                         NaN  \
1                                         NaN   
2                                         NaN   
3                                         NaN   
4                                         NaN   

  Number of full-time non-teaching staff in total   
0                                             NaN  \
1                                             NaN   
2                                             NaN   
3                                             NaN   
4                                             NaN   

   Number of full-time teaching staff of retirement age   
0                                                NaN     \
1                                                NaN      
2                                                NaN      
3                                                NaN      
4                                                NaN      

  Number of students in total Number of classes in total   
0                         NaN                        NaN  \
1                         NaN                        NaN   
2                         NaN                        NaN   
3                         NaN                        NaN   
4                         NaN                        NaN   

  Total number of students and children receiving preschool education   
0                                                NaN                   \
1                                                NaN                    
2                                                NaN                    
3                                                NaN                    
4                            

In [7]:
list(dataset.columns)

['outid',
 'sextypename',
 'regname',
 'areaname',
 'tername',
 'regtypename',
 'eoname',
 'eotypename',
 'eoregname',
 'eoareaname',
 'eotername',
 'eoparent',
 'ukrtest',
 'ukrteststatus',
 'ukrball100',
 'ukrptname',
 'ukrptregname',
 'ukrptareaname',
 'ukrpttername',
 'histtest',
 'histlang',
 'histteststatus',
 'histball100',
 'histptname',
 'histptregname',
 'histptareaname',
 'histpttername',
 'mathtest',
 'mathlang',
 'mathteststatus',
 'mathball100',
 'mathptname',
 'mathptregname',
 'mathptareaname',
 'mathpttername',
 'phystest',
 'physlang',
 'physteststatus',
 'physball100',
 'physptname',
 'physptregname',
 'physptareaname',
 'physpttername',
 'chemtest',
 'chemlang',
 'chemteststatus',
 'chemball100',
 'chemptname',
 'chemptregname',
 'chemptareaname',
 'chempttername',
 'biotest',
 'biolang',
 'bioteststatus',
 'bioball100',
 'bioptname',
 'bioptregname',
 'bioptareaname',
 'biopttername',
 'geotest',
 'geolang',
 'geoteststatus',
 'geoball100',
 'geoptname',
 'geoptreg

In [6]:
dataset.shape

(1967014, 128)

# Feature Description

EIE dataset contains 1,967,014 individuals and 128 features (general, test performance, and school information)

## General features (demographic, educational etc.)

`outid` – unique identifier of the participant

`sextypename` – gender of the participant ['Male', 'Female']

In [61]:
dataset.sextypename.value_counts(dropna=False)

sextypename
Female    994053
Male      972961
Name: count, dtype: int64

`regname` – region of registration/residence of the participant in Ukraine (24 oblasts and Kyiv, the capital of the country)

In [23]:
sorted(dataset.regname.unique())

['Cherkasy Oblast',
 'Chernihiv Oblast',
 'Chernivtsi Oblast',
 'Dnipropetrovsk Oblast',
 'Donetsk Oblast',
 'Ivano-Frankivsk Oblast',
 'Kharkiv Oblast',
 'Kherson Oblast',
 'Khmelnytskyi Oblast',
 'Kirovohrad Oblast',
 'Kyiv City',
 'Kyiv Oblast',
 'Luhansk Oblast',
 'Lviv Oblast',
 'Mykolaiv Oblast',
 'Odesa Oblast',
 'Poltava Oblast',
 'Rivne Oblast',
 'Sumy Oblast',
 'Ternopil Oblast',
 'Vinnytsia Oblast',
 'Volyn Oblast',
 'Zakarpattia Oblast',
 'Zaporizhzhia Oblast',
 'Zhytomyr Oblast']

`areaname` – district/city of registration/residence of the participant (a subdivision of regname)

`tername` – place of registration/residence of the participant (a subdivision of areaname)

`regtypename` – an educational status of the participant (5 types: 'A graduate of an Ukrainian school of the current year', 'A graduate of a school of the previous years', 'College student of a vocational or higher education institution', 'Penitentiary institution', 'A graduate of a foreign school')

In [29]:
dataset.regtypename.unique()

array(['A graduate of an Ukrainian school of the current year',
       'A graduate of a school of the previous years',
       'College student of a vocational or higher education institution',
       'Penitentiary institution', 'A graduate of a foreign school'],
      dtype=object)

In [59]:
pivot_df = dataset.groupby(['regtypename', 'year']).size().unstack()
pivot_df.loc['Total'] = pivot_df.sum()
pivot_df


year                                                    2016      2017   
regtypename                                                              
A graduate of a foreign school                          62.0      62.0  \
A graduate of a school of the previous years         48506.0   34424.0   
A graduate of an Ukrainian school of the curren...  200726.0  192785.0   
College student of a vocational or higher educa...   18325.0   13618.0   
Penitentiary institution                               384.0       NaN   
Total                                               268003.0  240889.0   

year                                                    2018      2019   
regtypename                                                              
A graduate of a foreign school                          32.0      33.0  \
A graduate of a school of the previous years         29307.0   33103.0   
A graduate of an Ukrainian school of the curren...  185373.0  187659.0   
College student of a vocational or higher educa...  120975.0  133018.0   
Penitentiary institution                                 NaN       NaN   
Total                                               335687.0  353813.0   

year                                                    2020      2021  
regtypename                                                             
A graduate of a foreign school                          45.0      36.0  
A graduate of a school of the previous years         42419.0   51009.0  
A graduate of an Ukrainian school of the curren...  211195.0  219640.0  
College student of a vocational or higher educa...  125640.0  118638.0  
Penitentiary institution                                 NaN       NaN  
Total                                               379299.0  389323.0

In 2018 it becames obligatory to take EIE for college students of a vocational or higher education institution to be admitted at the universtity.	

`tertypename` – type of territory (3 types: 'Village', 'City', 'Urban-type settlement')

In [30]:
dataset.tertypename.unique()

array(['Village', 'City', 'Urban-type settlement', nan], dtype=object)

In [63]:
pivot_df = dataset.groupby(['tertypename', 'year'], dropna=False).size().unstack()
pivot_df.loc['Total'] = pivot_df.sum()
pivot_df


year                       2016      2017      2018      2019      2020   
tertypename                                                               
City                   159036.0  152203.0  245246.0  263006.0  278221.0  \
Urban-type settlement   22075.0   21465.0   27032.0   27775.0   30326.0   
Village                 86891.0   67221.0   63409.0   63032.0   70752.0   
NaN                         1.0       NaN       NaN       NaN       NaN   
Total                  268003.0  240889.0  335687.0  353813.0  379299.0   

year                       2021  
tertypename                      
City                   284190.0  
Urban-type settlement   31515.0  
Village                 73618.0  
NaN                         NaN  
Total                  389323.0

In [32]:
dataset[dataset.tertypename.isnull()]

outid sextypename   
144704  3b07c9d4-067c-47fa-8cf4-8e2097ade009        Male  \

                      regname               areaname                 tername   
144704  Dnipropetrovsk Oblast  Dnipropetrovsk region  Petropavlovsk district  \

                                              regtypename eoname eotypename   
144704  College student of a vocational or higher educ...    NaN        NaN  \

       eoregname eoareaname  ... ukradaptscale Degree of institution   
144704       NaN        NaN  ...           NaN                   NaN  \

       Number of full-time teaching staff in total   
144704                                         NaN  \

       Number of full-time non-teaching staff in total   
144704                                             NaN  \

        Number of full-time teaching staff of retirement age   
144704                                                NaN     \

       Number of students in total Number of classes in total   
144704                         NaN                        NaN  \

       Total number of students and children receiving preschool education   
144704                                                NaN                   \

       Expenditures for the operation of the institution (general fund), UAH thousand (cash expenditures)   
144704                                                NaN                                                  \

       Expenditures for the operation of the institution (general fund), UAH thousand (plan)  
144704                                                NaN                                     

[1 rows x 128 columns]

`classprofilename`– profile of the participant's training class (24 types and two specific types: 'Junior specialit' for students and 'Skilled worker' for graduated people). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

In [33]:
dataset.classprofilename.unique()

array([nan, 'Physics and mathematics', 'History', 'Universal',
       'Ukrainian philology', 'Foreign philology',
       'Information technology', 'Other (multi-profile)', 'Mathematics',
       'Ecology', 'Economics', 'Technology', 'Geography',
       'Biology and chemistry', 'Sports', 'Physics',
       'Military and sports', 'Art and aesthetics', 'Law',
       'Biotechnology', 'Chemical technologies and agrochemistry',
       'Biology and physics', 'Philosophy', 'Physics and chemistry',
       'Інший(багатопрофільність)', 'Biology', 'Junior specialist',
       'Skilled worker'], dtype=object)

In [48]:
dataset[dataset.classprofilename.isnull()][['regtypename', 'year']]

regtypename  year
0        A graduate of an Ukrainian school of the curre...  2016
1        A graduate of an Ukrainian school of the curre...  2016
2             A graduate of a school of the previous years  2016
3        A graduate of an Ukrainian school of the curre...  2016
4        College student of a vocational or higher educ...  2016
...                                                    ...   ...
1966598       A graduate of a school of the previous years  2021
1966697       A graduate of a school of the previous years  2021
1966698       A graduate of a school of the previous years  2021
1966699       A graduate of a school of the previous years  2021
1966832       A graduate of a school of the previous years  2021

[594944 rows x 2 columns]

In [53]:
dataset[dataset.classprofilename.isnull()].groupby(['regtypename', 'year']).size().unstack()


year                                                    2016     2017   
regtypename                                                             
A graduate of a foreign school                          62.0     62.0  \
A graduate of a school of the previous years         48506.0  34424.0   
A graduate of an Ukrainian school of the curren...  200726.0      NaN   
College student of a vocational or higher educa...   18325.0  13618.0   
Penitentiary institution                               384.0      NaN   

year                                                    2018     2019   
regtypename                                                             
A graduate of a foreign school                          32.0     33.0  \
A graduate of a school of the previous years         29307.0  33103.0   
A graduate of an Ukrainian school of the curren...       NaN      NaN   
College student of a vocational or higher educa...  120975.0    989.0   
Penitentiary institution                                 NaN      NaN   

year                                                   2020     2021  
regtypename                                                           
A graduate of a foreign school                         45.0     36.0  
A graduate of a school of the previous years        42419.0  51009.0  
A graduate of an Ukrainian school of the curren...      NaN      NaN  
College student of a vocational or higher educa...    352.0    537.0  
Penitentiary institution                                NaN      NaN

`classprofilename`– profile of the participant's training class (24 types and two specific types: 'Junior specialit' for students and 'Skilled worker' for graduated people). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`classlangname` – participant's language of study (6 types and 'other' category). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eoname` – the name of participant's institution of education. Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eotypename` – type of educational institution (30 types: 'Lyceum', 'Gymnasium', etc). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eoregname` – region where the educational institution is located (same list as in regname). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eoareaname` – district/city where the educational institution is located (same list as areaname). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eotername` – place where the educational institution is located (same list as tername). Full information for the graduates of a Ukrainian school of the current year, partially for college graduates and students of a higher education institution. For else the values are NaN.

`eoparent` – the authority to which the educational institution is subordinated. Full information for the graduates of a Ukrainian school of the current year (except 2021), partially for college graduates and students of a higher education institution. For else the values are NaN.

`year` - individual's participation year

`age` - individual's age at the moment of the participation

## Test performance features 

There are several subjects to choose from:
- ukr (ukrainian language and literature)
- hist (Ukrainian history)
- math (Mathematics)
- phys (Physics)
- chem (Chemistry)
- bio (biology)
- geo (geography)
- eng (English language)
- fr (French language)
- deu (Deutch language)
- sp (Spanish language)

We will decode one of them as `subj` instead of one of the 11 subjects mentioned above.

So, for each subject we have the following features:

`subjtest` - the name of the subject (corresponding name or nan if the participant didn't choose to take it)

`subjteststatus` – the result of the test ('The score threshold wasn't reached', 'Test passed', 'No attendance',
 'Test results were cancelled', '100-200 wasn't selected', or NaN if the participant didn't choose to take it)
       
`subjball`– raw score of the corresponding subject for participants with statuses: 'Test passed' and 'The score threshold wasn't reached' by subjteststatus (for else is NaN).

`subjptname`– the place of conducting the ZNO (corresponding name or nan if the participant didn't choose to take it)

`subjptregname` – the region where the ZNO is conducted (corresponding name, similar to regname, or nan if the participant didn't choose to take it)

`subjptareaname` – district/city where the ZNO testing center is located (corresponding name, similar to areaname, or nan if the participant didn't choose to take it)

`subjpttername` – place where the ZNO testing center is located (corresponding name, similar to tername, or nan if the participant didn't choose to take it)

In [8]:
for cols in dataset.columns:
    print(cols, ':', end=' ')
    if len(dataset[cols].unique())>6:
        print('{ number:', len(dataset[cols].unique()), ', nan:', dataset[cols].isnull().sum(),'}')
    else:
        print(dataset[cols].unique(), ', nan:', dataset[cols].isnull().sum())

outid : { number: 1967014 , nan: 0 }
sextypename : ['Male' 'Female'] , nan: 0
regname : { number: 25 , nan: 0 }
areaname : { number: 590 , nan: 0 }
tername : { number: 17513 , nan: 0 }
regtypename : ['A graduate of an Ukrainian school of the current year'
 'A graduate of a school of the previous years'
 'College student of a vocational or higher education institution'
 'Penitentiary institution' 'A graduate of a foreign school'] , nan: 0
eoname : { number: 22518 , nan: 275623 }
eotypename : { number: 30 , nan: 275623 }
eoregname : { number: 26 , nan: 275623 }
eoareaname : { number: 747 , nan: 275623 }
eotername : { number: 6533 , nan: 275623 }
eoparent : { number: 2828 , nan: 279285 }
ukrtest : ['Ukrainian language and literature' nan] , nan: 157289
ukrteststatus : ['Test passed' 'No attendance' "The score threshold wasn't reached" nan
 'Test results were cancelled' "100-200 wasn't selected"] , nan: 157289
ukrball100 : { number: 111 , nan: 324904 }
ukrptname : { number: 2895 , nan: 157

In [9]:
dataset['eotypename'].value_counts()

eotypename
Secondary general education school                                677278
Higher educational institution                                    185117
Educational complex                                               160215
Lyceum                                                            133145
Gymnasium                                                          97332
Specialized school                                                 94686
Higher educational institution of I-II levels of accreditation     75988
Professional lyceum of the relevant profile                        48767
Institution of vocational education                                48714
Institution of professional preliminary higher education           46341
Higher vocational school                                           41481
Vocational school of the relevant profile                          24808
Center of professional and technical education                     12987
Educational association                 

In [10]:
dataset[dataset['classlangname'].isnull()].regtypename.unique()

array(['A graduate of an Ukrainian school of the current year',
       'A graduate of a school of the previous years',
       'College student of a vocational or higher education institution',
       'Penitentiary institution', 'A graduate of a foreign school'],
      dtype=object)

In [11]:
dataset[dataset['classlangname'].notnull()].regtypename.unique()

array(['A graduate of an Ukrainian school of the current year',
       'College student of a vocational or higher education institution'],
      dtype=object)

In [12]:
dataset['classprofilename'].unique()

array([nan, 'Physics and mathematics', 'History', 'Universal',
       'Ukrainian philology', 'Foreign philology',
       'Information technology', 'Other (multi-profile)', 'Mathematics',
       'Ecology', 'Economics', 'Technology', 'Geography',
       'Biology and chemistry', 'Sports', 'Physics',
       'Military and sports', 'Art and aesthetics', 'Law',
       'Biotechnology', 'Chemical technologies and agrochemistry',
       'Biology and physics', 'Philosophy', 'Physics and chemistry',
       'Інший(багатопрофільність)', 'Biology', 'Junior specialist',
       'Skilled worker'], dtype=object)

In [13]:
dataset.classprofilename.value_counts()

classprofilename
Universal                                  239443
Junior specialist                          227506
Ukrainian philology                        218325
Skilled worker                             147912
Foreign philology                          112719
Mathematics                                 77652
Technology                                  57675
History                                     50561
Інший(багатопрофільність)                   37815
Information technology                      35810
Other (multi-profile)                       24024
Biology and chemistry                       23581
Physics and mathematics                     23570
Economics                                   17478
Law                                         15398
Military and sports                         11990
Sports                                      11183
Biotechnology                                8651
Ecology                                      7443
Biology                          

In [14]:
dataset.classlangname.value_counts()

classlangname
Ukrainian    1289672
Russian        72454
Romanian        4915
Hungarian       3644
Moldovan         832
Polish           461
Other             92
Name: count, dtype: int64